### ISSUE TITLE
PyQt 5.13.2 app does not draw window content

### ISSUE THREAD

**Issue by ronaldoussoren** [ronaldoussoren/py2app#271]
**[Original report](https://bitbucket.org/ronaldoussoren/py2app/issue/271) by Marko Luther (Bitbucket: [MAKOMO](https://bitbucket.org/MAKOMO), GitHub: [MAKOMO](https://github.com/MAKOMO)).**

----------------------------------------

My app bundle reports

_It does not make sense to draw an image when \[NSGraphicsContext currentContext\] is nil.  This is a programming error. Break on void \_NSWarnForDrawingImageWithNoCurrentContext\(void\) to debug.  This will be logged only once.  This may break in the future._

on starting up and then does only show window borders without content \(transparent canvas\) with PyQt5.13.2 installed. Builds done after downgrading to PyQt 5.13.1 work as expected.

**Comment by MAKOMO** [comment]
Issue has by now observed by others also on pyinstaller https://github.com/pyinstaller/pyinstaller/issues/4627. It remains unclear what the underlying cause is. Qt, PyQt, Python, the bundler!?

**Comment by ronaldoussoren** [comment]
My current plan is to investigate a number of PyQt issues next weekend.

**Comment by MAKOMO** [comment]
You can use any PyQt app like the following simple one to reproduce this:

```
import sys
from PyQt5 import QtCore, QtWidgets
from PyQt5.QtWidgets import QMainWindow, QLabel, QGridLayout, QWidget
from PyQt5.QtCore import QSize

from PyQt5 import sip

class HelloWindow(QMainWindow):
    def __init__(self, parent = None):
        super(HelloWindow, self).__init__(parent)

        self.setMinimumSize(QSize(640, 480))
        self.setWindowTitle("Hello")

        title = QLabel("Hello World from PyQt", self) 
        title.setAlignment(QtCore.Qt.AlignCenter) 

        self.setCentralWidget(title)

if __name__ == "__main__":
    app = QtWidgets.QApplication(sys.argv)
    mainWin = HelloWindow()
    mainWin.show()
    sys.exit(app.exec_())
```

Now if you put

```
	<key>NSRequiresAquaSystemAppearance</key>
	<string>false</string>
```

into the apps Info.plist to switch the automatic dark mode support of Qt on, the Window remains empty. Using PyQt 5.13.1 instead, the Window is displaying its content and the title bar shows the dark theme (if dark mode is turned on in macOS) as expected. With all PyQt versions later, like 5.13.2, 5.14.0 or 5.14.1 the window remains empty and the console prints that strange macOS error message.

**Comment by ronaldoussoren** [comment]
This is almost certainly not a py2app issue, I can reproduce the issue as follows:

1) Edit ``/Library/Frameworks/Python.framework/Versions/3.8/Resources/Python.app/Contents/Info.plist`` to add the ``NSRequiresAquaSystemAppearance`` fragment

2) Run your script from the command-line (without using py2app). I get a dark window without text when the system is in dark mode, with the same error message you mention before.

This is with the Python.org installer for Python 3.8.1.

**Comment by MAKOMO** [comment]
I agree. Most likely a Qt or PyQt issue.

**Comment by ronaldoussoren** [comment]
I've created an app bundle for one of the PyObjC examples, which launches fine in light mode (even though my system is running in dark mode). When I add the ``NSRequiresAquaSystemAppearance`` key to ``Info.plist`` the app launches without issues and in dark mode. 

A different example does have problems in dark mode, but without the error message. I'm fairly sure those problems are caused by the NIB file used. 

Do you known which version of Xcode is used to build the PyQt5 packages?

**Comment by MAKOMO** [comment]
Sorry, no idea about the build environment of those PyQt5 wheels nor an idea how to find out. Thanks for digging into this. Quite a blogger for some apps.

**Comment by MAKOMO** [comment]
It turns out that with newer PyQt/Qt versions (>5.13.1) adding `NSRequiresAquaSystemAppearance ` to Info.plist is not needed (or even problematic) to get the app switching automatically to Qt's dark mode theme. Instead it is necessary that the app is linked to the macOS 10.14 or later SDK. 

However, the default apptemplates of py2app all link to the 10.8 SDK (while linking to 10.9 is required for notarization anyhow; see Issue #285). To fix this, I replace the definition of gPreBuildVariants in py2app/apptemplate/setup.py by the following

```
gPreBuildVariants = [
    {
        "name": "main-x86_64",
        "target": "10.14",
        "cflags": "-isysroot @@XCODE_ROOT@@/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk -arch x86_64",
        "cc": "/usr/bin/clang",
    },       
]
```

To recompile I had to remove the existing prebuilt templates first and then build the new one using setup.py.

```
# cd py2app/apptemplate
# rm -rf prebuilt
# python3 setup.py build
```

One needs to have XCode and the Command Line Tools installed for this to succeed. Then I installed py2app from this source directory via setup.py

```
# cd py2app
# python3 setup.py build
# python3 setup.py install
```

Building a Qt/PyQt 5.14.2 app with this "patched" py2app installation works as expected in dark mode as it links no to a corresponding macOS SDK version that supports the dark mode.

```
# otool -l Test.app/Contents/MacOS/Test | grep -A4 LC_BUILD_VERSION
LD_VERSION
       cmd LC_BUILD_VERSION
   cmdsize 32
  platform 1
       sdk 10.15.4
     minos 10.14
```

Not sure how py2app could support PyQt apps using dark mode and notarization out of the box.

**Comment by MAKOMO** [comment]
Related to Issue #285

**Comment by albertosottile** [comment]
I can confirm that I have the same issue with https://github.com/Syncplay/syncplay and PySide2. Rebuilding the prebuilt templates as suggested worked perfectly. However, a more permanent solution from the py2app side would probably be more practical.

**Comment by ronaldoussoren** [comment]
Fixing this will take some time, I'll have to perform some tests on older macOS releases after rebuilding the stub. 

At first I thought that #281 conflicted with this request, but that likely wasn't caused by the stub executable at all (the fix was a revert of a change to the Info.plist template).  But I still need to test that a rebuild of the stub doesn't reintroduce the problem mentioned in #281.

I don't know when I'll work on this, I'm currently at the (virtual) core dev sprint (working on finishing a PR for Python on Apple Silicon), after that I'll have to focus on PyObjC for a while to try to get a release out before macOS 11 goes GM.

**Comment by ronaldoussoren** [comment]
Hmmm, this is weird...


I've tested with both the reproducer app earlier in this discussion, as well as the "hello" example in the py2app repository, both with a setup.py file that adds the Info.plist key mentioned earlier to support dark mode.

Tested with Python 3.9 (python.org installer, "Intel" variant), and the current version of PyQt5 ("pip install pyqt5"). Testing was done in a VM running macOS 11 (Big Sur). 

The application starts file and does look correctly, both in light and dark mode.

The info.plist key is required to get dark mode support, without the key dark mode does not work.

Rebuilding the app stub removes the need to have the info.plist key.   

But annoyingly enough, that breaks Tkinter: With a rebuild stub I get a black window for the Tkinter "hello" example, both in light and dark mode (in the same setup as above).  That's likely due to the included Tkinter being pretty old.

The Tkinter issue is not present when I build using the universal2 installer on my M1 MacBook (Python.org installer, universal2 variant).  That includes a newer copy of Tkinter with loads of bug fixes, and the executable stub from py2app was (obviously) compiled using a recent SDK. 

I haven't tested PyQt5 with the universal2 Python on my laptop because PyQt5 does not have wheels supporting arm64.  Likewise for PyQt6 and PySide (2 and 6). I'm interested in trying to build all of PyQt and PyQt myself.

If anyone still has problems with PyQt and black windows I need more information on your setup:

* Which version of Python?
* How is Python installed (Python.org installer, home-brew, anaconda, ...)?
* Which version of macOS?
* Which CPU architecture(s) (x86_64, arm64, ...)?
* Which version of PyQt/PySide/....?
* How was that library installed?

Anything that might help in reproducing the issue on my end.
